tools/clang/include/clang/SPIRV/SpirvInstruction.h

```diff
@@ -14,277 +14,126 @@
 #include "clang/Basic/SourceLocation.h"
 #include "llvm/ADT/Optional.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Casting.h"
 
 namespace clang {
 namespace spirv {
 
 /// \brief The base class for representing SPIR-V instructions.
 class SpirvInstruction {
 public:
+  enum Kind {
+    // "Metadata" kinds
+    // In the order of logical layout
+
+    IK_Capability,      // OpCapability
+    IK_Extension,       // OpExtension
+    IK_ExtInstImport,   // OpExtInstImport
+    IK_MemoryModel,     // OpMemoryModel
+    IK_EntryPoint,      // OpEntryPoint
+    IK_ExecutionMode,   // OpExecutionMode
+    IK_String,          // OpString (debug)
+    IK_Source,          // OpSource (debug)
+    IK_Name,            // Op*Name (debug)
+    IK_ModuleProcessed, // OpModuleProcessed (debug)
+    IK_Decoration,      // Op*Decorate
+    IK_Type,            // OpType*
+    IK_Constant,        // OpConstant*
+    IK_Variable,        // OpVariable
+
+    // Function structure kinds
+
+    IK_FunctionParameter, // OpFunctionParameter
+
+    IK_LoopMerge,      // OpLoopMerge
+    IK_SelectionMerge, // OpSelectionMerge
+
+    IK_Branch,            // OpBranch
```

Comment:
I think it is worth to add a comment here that the order of these enums matters (should change carefully)
e.g. our code does >=IK_Branch && <= IK_Unreachable.

Reply:
Good call. Done